## Numpy assignment solution

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/gawron/python-for-social-science/blob/master/numpy/full_numpy_assignment_solution.ipynb)

## 1 Preliminaries

In [1]:
# The usual preamble
import numpy as np

## Numpy

In [2]:
import numpy as np
x = np.array([2,3,1,0])
y =np.arange(10)
z = 2 + x

In [5]:
x,z

(array([2, 3, 1, 0]), array([4, 5, 3, 2]))

Write expressions in the next cell to retrieve 1 from `x`, 3 from `y`, and 3 from `z`.

In [3]:
print(x[2])
print(y[3])
print(z[2])

1
3
3


In the next cell, write an expression that generates a 5 by 5 array filled with zeros, and another  that generates a 1 by 1 array with a 1 in it.

In [6]:
np.zeros((5,5))

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [21]:
print(np.ones((1,1)))
np.ones((1,1))[0,0]

[[1.]]


1.0

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  

In the next cell, write an expression that uses an assignment to a splice to make all the negative values in `a` be 0.  Attention: This can be done more easily in numpy than it can in normal Python.  See if you can do it the easy way.

In [22]:
a = np.arange(10,-1,-1) - 3
print(a)
a[-3:] = 0
print(a)

[ 7  6  5  4  3  2  1  0 -1 -2 -3]
[7 6 5 4 3 2 1 0 0 0 0]


In [10]:
a

array([7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0])

Or without peeking at the array:

In [11]:
a = np.arange(10,-1,-1) - 3
print(a)
a[a < 0] = 0
print(a)

[ 7  6  5  4  3  2  1  0 -1 -2 -3]
[7 6 5 4 3 2 1 0 0 0 0]


Another option is to `numpy` `where`, combined with an assignment.
The form is

```
>>> np.where(condition, x, y)
```

where `condition`, `x` and `y` are all arrays of the same size (or are broadcastable to the same size).
Return an array containing the members of `x` for which `condition` is `True`; otherwise return the corresponding member  of `y`.  It can be read "`x` where `condition`, else
`y`".

```
>>> a = np.array([1, 4.1, 9, 16])
>>> b = np.array([1,2,3,4])
>>> np.where((np.sqrt(a) - b) == 0,np.sqrt(a),b)
array([1., 2., 3., 4.])
```

In [12]:
a = np.arange(10,-1,-1) - 3
a = np.where(a < 0, 0, a)

In [13]:
a

array([7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0])

In the next cell write an expression that produces an array containing result of multiplying 7 by each of the first 12 integers (1 - 12). There's a hard way to do this and an easy way.  The easy way uses elementwise operations.

In [13]:
7 * np.arange(1,13)

array([ 7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84])

## 2D arrays

In [14]:
import numpy as np
room_matrix = \
np.array(
[[6,  3, 4, 1],
[5,  2, 3, 2],
[8,  3, 6, 2],
[5,  1, 3, 1],
[10, 4, 7, 2]])

cost_vector = np.array([40, 175, 90, 450])

In the next cell write an express that retrieves the last column of `room_matrix`.

In [15]:
room_matrix[:,-1]

array([1, 2, 2, 1, 2])

In the next cell, write an expression that retrieves the third row of `room_matrix`.

In [16]:
room_matrix[2,:]

array([8, 3, 6, 2])

In [17]:
room_matrix[2]

array([8, 3, 6, 2])

In the next cell write an expression that retrieves the first two elements of the second column of `room_matrix`

In [16]:
room_matrix[:2,1]

array([3, 2])

In the next cell write an expression that retrieves the following submatrix from `room_matrix`:

```
array([[2,3],
       [3,6]])
```

In [18]:
room_matrix[1:3,1:3]

array([[2, 3],
       [3, 6]])

In the next cell, write an expression that returns a Boolean array that identifies the members of `room_matrix` which are greater than or equal to 7.

In [19]:
room_matrix >= 7

array([[False, False, False, False],
       [False, False, False, False],
       [ True, False, False, False],
       [False, False, False, False],
       [ True, False,  True, False]])

In the next cell, write an expression that produces an array containing the items in `room_matrix` which are greater than or equal to 7.  Note:  This question and the previous question do not have the same answer.

In [20]:
room_matrix[room_matrix >= 7]

array([ 8, 10,  7])

## Iris data exercise

In the next cell, use a Boolean mask to return all the rows of the iris data set whose column 3 value is greater than 1.8 (I mean the column whose **index** is 3, the last column).
The cell is not empty because you have been provided with the code which loads the data.
So you should keep that code and **add to it**.

In [1]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']
features[features[:,3] > 1.8,:]

array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.7, 3.3, 5.7, 2.1],
       [6.4, 2.8, 5.6, 2.1],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [7.7, 3. , 6.1, 2.3],
       [6.3, 3.4, 5.6, 2.4],
       [6.9, 3.1, 5.4, 2.1],
       [6.7, 3.1, 5.6, 2.4],
       [6.9, 3.1, 5.1, 2.3],
       [5.8, 2.7, 5.1, 1.9],
       [6.8, 3.2, 5.9, 2.3],
       [6.7, 3.3, 5.7, 2.5],
       [6.7, 3. , 5.2, 2.3],
       [6.3, 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3]])

In the next cell, use a Boolean array to return all the rows representing members of class 2 in the iris data set (the Boolean array is not the answer; it is used in retrieving the answer).

In [33]:
features[target==2,:]

array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.3, 2.9, 5.6, 1.8],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [4.9, 2.5, 4.5, 1.7],
       [7.3, 2.9, 6.3, 1.8],
       [6.7, 2.5, 5.8, 1.8],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [6.5, 3. , 5.5, 1.8],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6. , 2.2, 5. , 1.5],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.3, 2.7, 4.9, 1.8],
       [6.7, 3.3, 5.7, 2.1],
       [7.2, 3.2, 6. , 1.8],
       [6.2, 2.8, 4.8, 1.8],
       [6.1, 3. , 4.9, 1.8],
       [6.4, 2.8, 5.6, 2.1],
       [7.2, 3. , 5.8, 1.6],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [6.3, 2.8, 5.1, 1.5],
       [6.1, 2

Extra credit:  In the next cell write an expression -- or expressions -- which returns
all the members of class 2 in the data whose column 3 value is greater than 1.8.

Hint: Set a variable to the result of one of your answers to the last two questions.
That result is an array, and you can do more indexing on it.

Two Solutions.  The one not hinted at uses Boolean array conjunction.

In [4]:
# Doesnt work to do this part first.  Why not?
#XX = features[features[:,3] > 1.8,:]
XX = features[target==2,:]
print(XX[XX[:,3] > 1.8,:])
print(len(XX[XX[:,3] > 1.8,:]))

[[6.3 3.3 6.  2.5]
 [5.8 2.7 5.1 1.9]
 [7.1 3.  5.9 2.1]
 [6.5 3.  5.8 2.2]
 [7.6 3.  6.6 2.1]
 [7.2 3.6 6.1 2.5]
 [6.5 3.2 5.1 2. ]
 [6.4 2.7 5.3 1.9]
 [6.8 3.  5.5 2.1]
 [5.7 2.5 5.  2. ]
 [5.8 2.8 5.1 2.4]
 [6.4 3.2 5.3 2.3]
 [7.7 3.8 6.7 2.2]
 [7.7 2.6 6.9 2.3]
 [6.9 3.2 5.7 2.3]
 [5.6 2.8 4.9 2. ]
 [7.7 2.8 6.7 2. ]
 [6.7 3.3 5.7 2.1]
 [6.4 2.8 5.6 2.1]
 [7.4 2.8 6.1 1.9]
 [7.9 3.8 6.4 2. ]
 [6.4 2.8 5.6 2.2]
 [7.7 3.  6.1 2.3]
 [6.3 3.4 5.6 2.4]
 [6.9 3.1 5.4 2.1]
 [6.7 3.1 5.6 2.4]
 [6.9 3.1 5.1 2.3]
 [5.8 2.7 5.1 1.9]
 [6.8 3.2 5.9 2.3]
 [6.7 3.3 5.7 2.5]
 [6.7 3.  5.2 2.3]
 [6.3 2.5 5.  1.9]
 [6.5 3.  5.2 2. ]
 [6.2 3.4 5.4 2.3]]
34


Solution 2.  This one uses `&`.

In [6]:
print(features[(features[:,3] > 1.8) & (target == 2)])
len(features[(features[:,3] > 1.8) & (target == 2)])

[[6.3 3.3 6.  2.5]
 [5.8 2.7 5.1 1.9]
 [7.1 3.  5.9 2.1]
 [6.5 3.  5.8 2.2]
 [7.6 3.  6.6 2.1]
 [7.2 3.6 6.1 2.5]
 [6.5 3.2 5.1 2. ]
 [6.4 2.7 5.3 1.9]
 [6.8 3.  5.5 2.1]
 [5.7 2.5 5.  2. ]
 [5.8 2.8 5.1 2.4]
 [6.4 3.2 5.3 2.3]
 [7.7 3.8 6.7 2.2]
 [7.7 2.6 6.9 2.3]
 [6.9 3.2 5.7 2.3]
 [5.6 2.8 4.9 2. ]
 [7.7 2.8 6.7 2. ]
 [6.7 3.3 5.7 2.1]
 [6.4 2.8 5.6 2.1]
 [7.4 2.8 6.1 1.9]
 [7.9 3.8 6.4 2. ]
 [6.4 2.8 5.6 2.2]
 [7.7 3.  6.1 2.3]
 [6.3 3.4 5.6 2.4]
 [6.9 3.1 5.4 2.1]
 [6.7 3.1 5.6 2.4]
 [6.9 3.1 5.1 2.3]
 [5.8 2.7 5.1 1.9]
 [6.8 3.2 5.9 2.3]
 [6.7 3.3 5.7 2.5]
 [6.7 3.  5.2 2.3]
 [6.3 2.5 5.  1.9]
 [6.5 3.  5.2 2. ]
 [6.2 3.4 5.4 2.3]]


34

In [36]:
len(XX)

34

## Boston Housing data exercise (UCI ML housing dataset)

In [4]:
from sklearn import datasets
boston = datasets.load_boston()
X,Y = boston.data,boston.target
print(X.shape, Y.shape)

(506, 13) (506,)


The cell above loads the `sklearn` Boston housing data, which contains information about housing prices
for 501 houses sold in Boston, and the values of various attributes which have been found to be useful
in predicting home prices:

```
Index
0         - CRIM     per capita crime rate by town
1         - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
2         - INDUS    proportion of non-retail business acres per town
3         - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
4         - NOX      nitric oxides concentration (parts per 10 million)
5         - RM       average number of rooms per dwelling
6         - AGE      proportion of owner-occupied units built prior to 1940
7         - DIS      weighted distances to five Boston employment centres
8         - RAD      index of accessibility to radial highways
9         - TAX      full-value property-tax rate per $10,000
10        - PTRATIO  pupil-teacher ratio by town
11        - B        1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
12        - LSTAT    % lower status of the population
```

The array `X`, defined in the cell above, is a 501x13 array containing the values of all
these variables for 501 Boston houses.  

The information stored in Y, the value to be predicted is

```
       - MEDV     Median value of owner-occupied homes in $1000's
 
```

1.  Construct an expression or expressions that creates a new table that omits all houses whose tracts
    bound the Charles River.
2.  Construct an expression or expressions that creates a new table containing only the CRIM column and
    the ZN column.  This is  our new X.  Also create a new Y containing the pupil-teach ratio.
    This is to help someone research how to predict poor education conditions.  The shape of the new X should
    (506,2) and the shape of the new Y should be (506,).  Show that you checked this in your answer.

In [39]:
NoCharles = X[X[:,3]==0,:]

In [40]:
len(X),len(NoCharles)

(506, 471)

How many 1s in the third column?

In [44]:
sum(X[:,3])

35.0

Checks!

## Part two

In [7]:
# On the CRIM and ZN Cols
XX = X[:,:2]
YY = X[:,10]
print(XX.shape)
print(YY.shape)

(506, 2)
(506,)


The answers for the following section are available on the inclass Boolean exercies solution notebook.

# Boolean combinations

In this section we load forest covertype data.  This is a multi class dataset that has data for 7 different forest covertypes (stored in the `target` attribute).  There are 581,012 forest plots with 54 attributes each (stored ina 581012x54 array) The first ten attributes are numerical, the last 44 are Boolean (true/false) attributes. Each of the Boolean attributes represents a qualitative soil type attribute which is either present or absent.  We will refer to all the attributes by their column index.  Thus the first attribute is attribute 0 and the last (a Boolean attribute) is attribute 53.

In [ ]:
#from sklearn.datasets import load_wine
#wdata = load_wine()
from sklearn.datasets import fetch_covtype
data = fetch_covtype()
print(data.data.shape)  # data.data is the 581012 x 54 array
print(target.shape)     # data.target contains the class for each instance

[The Boolean arrays and masks notebook](https://github.com/gawron/python-for-social-science/blob/master/numpy/02_06_Boolean_Arrays_and_Masks.ipynb) discusses combining Boolean arrays with Boolean operators `&` (conceptually 'and') and `|` (conceptually 'or').   We can add to that `~` (conceptually 'not') to complete the list.  Look at the examples there.

 This problem has two parts: 1. Construct a single Python expression which contains all the rows of the covertype dataset  that are not of class 3 or 5,
have a value greater than 300 for attribute 1, and have attribute 12 (a qualitative soil attribute). 2. Construct another single Python expression which counts the number of plots you found in part 1. 

The `sklearn` description of the covertype dataset is printed in the cell below. For a fuller understanding of the attributes in the dataset you might want to look at [the original UCI data set description.](https://archive.ics.uci.edu/ml/datasets/Covertype)

In [ ]:
print(data.DESCR)

.. _covtype_dataset:

Forest covertypes
-----------------

The samples in this dataset correspond to 30×30m patches of forest in the US,
collected for the task of predicting each patch's cover type,
i.e. the dominant species of tree.
There are seven covertypes, making this a multiclass classification problem.
Each sample has 54 features, described on the
`dataset's homepage <https://archive.ics.uci.edu/ml/datasets/Covertype>`__.
Some of the features are boolean indicators,
while others are discrete or continuous measurements.

**Data Set Characteristics:**

    =================   ============
    Classes                        7
    Samples total             581012
    Dimensionality                54
    Features                     int
    =================   ============

:func:`sklearn.datasets.fetch_covtype` will load the covertype dataset;
it returns a dictionary-like object
with the feature matrix in the ``data`` member
and the target values in ``target``.
The dataset will be d

`data.data` attribute is a `numpy` array containing the array.  As an example to help with this problem.  To compute an an array containing just the rows in which attribute 0 is greater than 50 you do.  Remember the

In [ ]:
WW = data.data > 50
print(WW.shape)

(581012, 54)

In [ ]:
#print(set(data.data[:,-44:].ravel()))
#print(
print(sum((data.data[:,1] > 300) & ~((data.target == 3) | (data.target == 5)) & (data.data[:,12] == 1)))
print(sum((data.data[:,1] > 300) &  (data.data[:,12] == 1)))
#data.data[:,12].shape
#((data.data[:,1] > 300) & (data.target == 3 | data.target == 5)).shape
#and import numpy as np
#np.bitwise_or(data.target == 3,data.target == 5).shape# | data.target == 5).shape
#((data.target == 3) | (data.target == 5)).shape
#print(data.data.shape)

43815
45076


The covertypes (or dominant tree) for each forest plot are in `data.target`, which is often the attribute used to store the classes in an `sklearn` clasification dataset.

In [ ]:
print(len(data.target))
print(set(data.target))

581012
{1, 2, 3, 4, 5, 6, 7}


** The solutions to these problem,s were posted as in class Boolean exercises solution notebook.**